EssentialCSharp.Web/Services/ReferralService.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System.Security.Claims;
+using EssentialCSharp.Web.Areas.Identity.Data;
+using EssentialCSharp.Web.Data;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.EntityFrameworkCore;
+using Sqids;
+
+namespace EssentialCSharp.Web.Services;
+
+public class ReferralService(EssentialCSharpWebContext dbContext, SqidsEncoder<int> sqids, UserManager<EssentialCSharpWebUser> userManager) : IReferralService
+{
+    public async Task<string?> GetReferralIdAsync(string userId)
+    {
+        EssentialCSharpWebUser? user = await userManager.FindByIdAsync(userId);
+        if (user is null)
+        {
+            return null;
+        }
+        else
+        {
+            // Check if the user already has a referrer ID
+            if (!string.IsNullOrEmpty(user.ReferrerId))
+            {
+                return user.ReferrerId;
+            }
+            else
+            {
+                Random random = new();
+                string referrerId = sqids.Encode(random.Next());
+                user.ReferrerId = referrerId;
+
+                await userManager.AddClaimAsync(user, new Claim("ReferrerId", referrerId));
+                await userManager.UpdateAsync(user);
+                return user.ReferrerId;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Track the referral in the database.
+    /// </summary>
+    /// <param name="referralId">The referrer ID to track.</param>
+    /// <returns>True if the referral was successfully tracked, otherwise false.</returns>
+    public async Task<bool> TrackReferralAsync(string referralId, ClaimsPrincipal? user)
+    {
+        EssentialCSharpWebUser? claimsUser = user is null ? null : await userManager.GetUserAsync(user);
+        if (claimsUser is null)
+        {
+            return await TrackReferral(dbContext, referralId);
+        }
+        else
+        {
+            // If the user is the referrer, do not track the referral
+            if (claimsUser.ReferrerId == referralId)
+            {
+                return false;
+            }
+            else
+            {
+                return await TrackReferral(dbContext, referralId);
+            }
+        }
+
+        static async Task<bool> TrackReferral(EssentialCSharpWebContext dbContext, string referralId)
+        {
+            EssentialCSharpWebUser? dbUser = await dbContext.Users.SingleOrDefaultAsync(u => u.ReferrerId == referralId);
+            if (dbUser is null)
+            {
+                return false;
+            }
+            else
+            {
+                dbUser.ReferralCount++;
```

Comment:
Probably a lot more than what we need for v1, but this looks like a concurrency issue waiting to happen.
https://learn.microsoft.com/en-us/ef/core/saving/concurrency?tabs=data-annotations&WT.mc_id=DT-MVP-5003472

Reply:
Addressed in https://github.com/IntelliTect/EssentialCSharp.Web/pull/722/commits/ebb084728bdbe5e78accff9235e8fadd589392eb

I couldn't cause the exception to be thrown to test

Reply:
For an update so simple as this, the much, much simpler way to address this would be to use `ExecuteUpdate`. Let the database do the hard work rather than re-writing ACID compliance in your application code.

`db.Users.Where(u => u.ReferrerId == referralId).ExecuteUpdateAsync(setters => setters.SetProperty(b => b.ReferralCount, b => b.ReferralCount+ 1));`

Reply:
I haven't used this before! Thanks, much simpler, and interesting all the different ways to use that when reading :) 

https://learn.microsoft.com/ef/core/saving/execute-insert-update-delete?WT.mc_id=8B97120A00B57354#referencing-the-existing-property-value